lib/Sema/LegalConstExprVerifier.cpp

```diff
@@ -237,6 +233,63 @@ checkSupportedWithSectionAttribute(const Expr *expr,
       return std::make_pair(expr, OpaqueDeclRef);
     }
 
+    // Allow specific patterns of AutoClosureExpr, which is used in static func
+    // references. E.g. "MyStruct.staticFunc" is:
+    // - autoclosure_expr type="() -> ()"
+    //   - call_expr type="()"
+    //     - dot_syntax_call_expr
+    //       - declref_expr decl="MyStruct.staticFunc"
+    //       - dot_self_expr type="MyStruct.Type"
+    //         - type_expr type="MyStruct.Type"
+    if (const AutoClosureExpr *autoClosureExpr =
+            dyn_cast<AutoClosureExpr>(expr)) {
+      if (auto callExpr =
```

Comment:
I think `AutoClosureExpr::getUnwrappedCurryThunkExpr()` can help do a bunch of this digging for you.